src/gensxg.c

```diff
@@ -0,0 +1,444 @@
+  struct tm date;
+  char* finished = strptime(datestring, "%Y-%m-%dT%H:%M:%SZ", &date);
+  if (finished == NULL) {
+    fprintf(stderr, "Failed to parse date: %s\n", datestring);
+  } else if(*finished != '\0') {
+    fprintf(stderr, "Failed to parse at: %s\n", finished);
+    exit(EXIT_FAILURE);
+  }
+  return mktime(&date);
+}
+
+static Options init_default_options() {
+  Options result;
+  memset(&result, 0, sizeof(result));
+  result.content = "index.html";
+  result.content_type = "text/html";
+  result.header = sxg_empty_header();
+  result.date = time(NULL); /* = Now. */
+  result.duration = 60 * 60 * 24 * 7;  /* = 7 Days. */
+  result.mi_record_size = 4096;
+  return result;
+}
+
+static void header_append_string(const char* data, sxg_header_t* header) {
+  size_t colon_pos = 0;
+  for (size_t i = 0; data[i] != '\0'; ++i) {
+    if (data[i] == ':') {
+      colon_pos = i;
+      break;
+    }
+  }
+  if (colon_pos == 0) {
+    fprintf(stderr, "Invalid header format: %s\n", data);
+    exit(EXIT_FAILURE);
+  }
+  size_t value_pos = colon_pos + 1;
+  while (data[value_pos] != '\0' && data[value_pos] == ' ') ++value_pos;
+  char* const key = OPENSSL_strndup(data, colon_pos);
+  char* const value = OPENSSL_strdup(&data[value_pos]);
+
+  sxg_header_append_string(key, value, header);
+
+  OPENSSL_free(key);
+  OPENSSL_free(value);
+}
+
+static Options parse_options(int argc, char* const argv[]) {
+  Options result = init_default_options();
+  int opt;
+  int longindex;
+  time_t expires = 0;
+  bool is_expire_specified = false;
+  while ((opt = getopt_long_only(argc,
+                                 argv,
+                                 "h:i:c:t:H:m:u:r:v:e:p:k:d:x",
+                                 kOptions,
+                                 &longindex)) != -1) {
+    switch (opt) {
+      case 'h':
+        result.help = true;
+        return result;
+      case 'c':
+        result.content = optarg;
+        break;
+      case 't':
+        result.content_type = optarg;
+        break;
+      case 'H':
+        header_append_string(optarg, &result.header);
+        break;
+      case 'm':
+        result.mi_record_size = strtoull(optarg, NULL, 0);
+        break;
+      case 'u':
+        result.url = optarg;
+        break;
+      case 'r':
+        result.cert_url = optarg;
+        break;
+      case 'v':
+        result.validity_url = optarg;
+        break;
+      case 'e':
+        result.certificate = optarg;
+        break;
+      case 'p':
+        result.public_key = optarg;
+        break;
+      case 'k':
+        result.private_key = optarg;
+        break;
+      case 'd':
+        result.date = parse_time(optarg);
+        break;
+      case 'x':
+        is_expire_specified = true;
+        expires = parse_time(optarg);
+        break;
+      default:
+        fprintf(stderr, "unknown option %c\n", opt);
+        exit(EXIT_FAILURE);
+    }
+  }
+
+  /* Set default expiration time if not specified. */
```

Comment:
I think you're setting _non-default_ expiration time.

Reply:
Removed redundant comment.